Libraries/D3D12RaytracingFallback/src/TreeletReorder.hlsl

```diff
@@ -56,210 +45,289 @@ AABB CombineAABB(AABB aabb0, AABB aabb1)
+                if (BIT(i) & treeletBitmask)
                 {
-                    if (countbits(treeletBitmask) == bits)
-                    {
-                        float lowestCost = FLT_MAX;
-                        uint bestPartition = 0;
-                        for (uint partitionBitmask = 1; partitionBitmask < treeletBitmask; partitionBitmask++)
-                        {
-                            if ((treeletBitmask | partitionBitmask) == treeletBitmask)
-                            {
-                                const float cost = optimalCost[partitionBitmask] + optimalCost[treeletBitmask ^ partitionBitmask];
-                                if (cost < lowestCost)
-                                {
-                                    lowestCost = cost;
-                                    bestPartition = partitionBitmask;
-                                }
-                            }
-                        }
-                        // TODO: Can the CostOfRayBoxIntersection be baked into area[s]?
-                        optimalCost[treeletBitmask] = CostOfRayBoxIntersection * area[treeletBitmask] + lowestCost; // TODO: Consider cost of flattening to triangle list
-                        optimalPartition[treeletBitmask] = bestPartition;
-                    }
+                    aabb = CombineAABB(aabb, AABBBuffer[treeletToReorder[i]]);
                 }
             }
 
-            // Now that a reordering has been calculated, reform the tree
-            struct PartitionEntry
-            {
-                uint Mask;
-                uint NodeIndex;
-            };
-            uint nodesAllocated = 1;
-            uint partitionStackSize = 1;
-            PartitionEntry partitionStack[MaxTreeletSize];
-            partitionStack[0].Mask = FullPartitionMask;
-            partitionStack[0].NodeIndex = internalNodes[0];
-            while (partitionStackSize > 0)
+            // Intermediate value.
+            optimalCost[treeletBitmask] = ComputeBoxSurfaceArea(aabb);
+        }
+    }
+
+    AABB nodeAABB = AABBBuffer[nodeIndex];
+    float rootAABBSurfaceArea = ComputeBoxSurfaceArea(nodeAABB);
+
+    if (threadId < FullTreeletSize)
+    {
+        optimalCost[BIT(threadId)] = CalculateCost(AABBBuffer[treeletToReorder[threadId]], rootAABBSurfaceArea);
+    }
+
+    GroupMemoryBarrierWithGroupSync();
+
+    [unroll]
+    for (uint subsetSize = 2; subsetSize <= FullTreeletSize; subsetSize++)
+    {
+        uint numTreeletBitmasks = FullTreeletSizeChoose[subsetSize];
+        if (threadId < numTreeletBitmasks)
+        {
+            numBitmasksPerThread = max(numTreeletBitmasks / NumThreadsInGroup, 1);
+            extraBitmasks = numTreeletBitmasks > NumThreadsInGroup ? numTreeletBitmasks % NumThreadsInGroup : 0;
+
+            bitmasksStart = numBitmasksPerThread * threadId + min(threadId, extraBitmasks);
+            bitmasksEnd = bitmasksStart + numBitmasksPerThread + (threadId < extraBitmasks ? 1 : 0);
```

Comment:
Rather than splitting these into blocks of numBitmasksPerThread, you might want to experiment with reading in at strided offsets. I can swing by and explain, but on GPUs you might be able to benefit here from memory coalescing (I don't have a handy link to point you to off hand, but you can google some descriptions about it)

Reply:
[I reimplemented the loop to use strided offsets](https://gist.github.com/rgerd/2cd389a6160d18dd8773620e99bbc116), but it actually seemed to make it slower